File: app/src/main/java/com/siavash/dualcamera/util/Toolbar.java
package com.siavash.dualcamera.util;

import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.Bitmap;
import android.os.Build;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageButton;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.siavash.dualcamera.R;

import butterknife.Bind;
import butterknife.ButterKnife;

/**
 * Customized toolbar
 * Created by sia on 8/26/15.
 */
public class Toolbar<T> extends RelativeLayout implements View.OnClickListener {

    @Bind(R.id.title) TextView title;
    @Bind(R.id.back_btn) ImageButton backButton;

    private OnClickListener mCallback;

    public Toolbar(Context context) {
        super(context);
        setUp(context);
    }

    public Toolbar(Context context, AttributeSet attrs) {
        super(context, attrs);
        setUp(context);
    }

    public Toolbar(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setUp(context);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP) public Toolbar(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        setUp(context);
    }

    public void setUp(Context context) {
        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        View view = inflater.inflate(R.layout.layout_toolbar, this, true);
        ButterKnife.bind(this, view);

        backButton.setOnClickListener(this);
    }

    public void setTitle(String title) {
        this.title.setText(title);
    }

    public void setCallback(T callback) {
        mCallback = (OnClickListener) callback;
    }

    public void setRightButtonResource(int resId) {
        backButton.setImageResource(resId);
    }

    public void setRightButtonBitmap(Bitmap bitmap) {
        backButton.setImageBitmap(bitmap);
    }

    @Override public void onClick(View v) {
        int id = v.getId();
        if (id == backButton.getId()) {
            mCallback.onClick();
        }
    }

    public interface OnClickListener {
        void onClick();
    }
}

File: app/src/main/kotlin/com/siavash/dualcamera/util/Toolbar.kt
package com.siavash.dualcamera.util

import android.content.Context
import android.graphics.Typeface
import android.support.v4.view.GravityCompat
import android.support.v4.widget.DrawerLayout
import android.support.v7.widget.Toolbar
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageButton
import android.widget.TextView
import com.siavash.dualcamera.R
import org.jetbrains.anko.layoutInflater
import org.jetbrains.anko.onClick
import java.util.jar.Attributes

/**
 * Created by sia on 11/16/15.
 */
class Toolbar : Toolbar {
    private val title: TextView by bindView(R.id.toolbar_title)
    val right: ImageButton by bindView(R.id.toolbar_right)
    val left: ImageButton by bindView(R.id.toolbar_left)

    constructor(context: Context) : super(context) {
        setUp(context)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        setUp(context)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        setUp(context)
    }

    private fun setUp(context: Context) {
        context.layoutInflater.inflate(R.layout.layout_toolbar, this, true)
    }

    fun setTitle(title: String) {
        this.title.text = title
    }

    fun setTitleStyle(typeface: Typeface) {
        title.typeface = typeface
    }

    fun defaultTitleStyle() {
        title.typeface = getFont(context, Font.AFSANEH)
    }

    fun setRightItemVisibility(visibility: Int) {
        right.visibility = visibility
    }

    inline fun setRightAction(crossinline function: () -> Unit) {
        if (right.visibility != VISIBLE) setRightItemVisibility(VISIBLE)
        right.onClick {
            function()
        }
    }

    inline fun setLeftAction(crossinline function: () -> Unit) {
        left.onClick {
            function()
        }
    }

    fun defaultHamburgerAction(drawer: DrawerLayout) {
        left.onClick {
            if (drawer.isDrawerOpen(GravityCompat.END)) {
                drawer.closeDrawer(GravityCompat.END)
            } else {
                drawer.openDrawer(GravityCompat.END)
            }
        }
    }
}